T2D-307 — Fetch taxonomy of a study from reference sequences

tskir opened the pull request:
### [**Diff link**](https://github.com/tskir/metadata-ws/compare/t2d-215-study-child-objects-release...t2d-307-study-taxonomy?expand=1)

Implemented fetching Study's taxonomy through `@JoinFormula`. The relation is now Study M..1 Taxonomy. This replaces the old `taxonomy` field which had to be filled out manually.

Feature tests had to be extensively updated/rewritten to reflect this change. They are now all passing. `StudySteps` had to be updated as well.

I also reformatted example tables in all feature tests for better readatbiliity; please let me know what you think.

This PR is created on top of #53. To see only changes related to this PR, use this link: https://github.com/tskir/metadata-ws/compare/t2d-215-study-child-objects-release...t2d-307-study-taxonomy?expand=1

tskir commented:
Implemented fetching the taxonomy from the reference sequences. The relation is now M..1, as we discussed. This replaces the old `taxonomy` field which had to be specified explicitly. Removed TaxonomyExtractor.

Rewrote/updated feature tests and `StudySteps` module. Reformatted the example tables. Ready for the second iteration of review.

tskir commented:
Rebased this on top of the recent changes in #47, #50 and #53. **Ready for the third iteration of review.**

[_Handy diff link which includes only changes from this PR_](https://github.com/tskir/metadata-ws/compare/t2d-215-study-child-objects-release...t2d-307-study-taxonomy?expand=1)

tskir commented:
Rebased this on top of the recent changes in #47, #50 and #53. **Ready for the fourth iteration of review.** (Not that the previous three iterations produced much comments though :-)

[_Diff link_](https://github.com/tskir/metadata-ws/compare/t2d-215-study-child-objects-release...t2d-307-study-taxonomy?expand=1)

iyangar commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Study.java:
```diff
@@ -84,9 +86,15 @@
     @JsonProperty
-    @ManyToOne(optional = false)
+    @ManyToOne
+    @JoinFormula(value = "(SELECT DISTINCT taxonomy.id FROM taxonomy " +
```
Which is the new test case covering this?

tskir replied:
Instead of implementing new test cases, I modified the two _existing_ tests which previously covered Study—Taxonomy relation, to test for the new logic. Both tests are in the `study.feature` file.

Test 1 is **“Scenario Outline: search various study by taxonomy name and id”.** It creates five taxonomies, three studies, and then links taxonomies to studies **indirectly** through associating Study with an Analysis, Analysis with a ReferenceSequence, and ReferenceSequence with a Taxonomy. Then we query Studies directly by Taxonomy to ensure that taxonomy fetching is working correctly.

Test 2 is **“Scenario Outline: search various public studies”.** It again creates a taxonomy and several studies and them links them **indirectly** through analyses and reference sequences. The taxonomy fetching is then tested in conjunction with other parameters, such as a release date.

iyangar commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Study.java:
```diff
@@ -84,9 +86,15 @@
     private LocalDate releaseDate;
 
-    @ApiModelProperty(position = 7, dataType = "java.lang.String", notes = "Url to a Taxonomy")
+    @ApiModelProperty(position = 7, dataType = "java.lang.String", notes = "URL to a Taxonomy")
```
`URL to taxonomy`. Class names start with Caps but message specifies it and not necessarily class name

tskir replied:
Good point, fixed

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/converter/ReferenceSequenceConverter.java:
```diff
@@ -0,0 +1,67 @@
+        return convertSaveAndReturnNewReferenceSequence(assemblyType, taxonomy);
+    }
+
+    private Taxonomy findExistingorSaveandReturnTaxonomy(AssemblyType assemblyType) {
```
This function is not related to this PR (as you can see the file `ReferenceSequenceConverter.java` is not present in the [diff link](https://github.com/tskir/metadata-ws/compare/t2d-215-study-child-objects-release...t2d-307-study-taxonomy?expand=1) that I provided). Rather it is related to #50. Anyway this method does not exist anymore after the most recent changes in that PR, so the problem is kind of solved :)

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -18,21 +18,207 @@
+        return referenceSequenceAccessions;
+    }
+
+    private String getAccessionFromReferenceSequenceType(ReferenceSequenceType referenceSequenceType){
```
Discussed & fixed in #50 [(link to discussion)](https://github.com/EBIvariation/metadata-ws/pull/50#pullrequestreview-242547661)

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java:
```diff
@@ -21,18 +21,10 @@
-import javax.persistence.SequenceGenerator;
-import javax.persistence.Table;
-import javax.persistence.UniqueConstraint;
+import javax.persistence.*;
```
Fixed in the last code version

tskir commented:
Rebased on top of recent changes in #50 and #53. Addressed all review comments. **Ready for the fifth iteration of review.**

[Diff link](https://github.com/tskir/metadata-ws/compare/t2d-215-study-child-objects-release...t2d-307-study-taxonomy?expand=1)

selvaebi commented on metadata-ws/src/test/resources/features/study.feature:
```diff
@@ -60,26 +90,30 @@ Feature: study object
-      | taxonomy-name | name=Pan | 2 | STUDY2,STUDY3 |
-      | taxonomy-name | name=Homininae | 3 | STUDY1,STUDY2,STUDY3 |
-      | taxonomy-name | name=None | 0 | NONE |
+      | base          | query             | N | url                  |
```
Can we have a test for direct search as well, like fetching the taxonomy of a given study ?

tskir replied:
Expanded the first scenario in `study.feature` file to fetch & check taxonomy of a study directly. This scenario is now called “register a study successfully and check its fields”.

tskir commented:
@cyenyxe @iyangar Rebased, fixed conflicts & addressed review comment. This PR is ready for review. Diff link is in the description.

selvaebi commented on metadata-ws/src/test/resources/features/study.feature:
```diff
@@ -1,22 +1,36 @@
-    When I request POST /taxonomies with JSON payload:
+  Scenario: register a study successfully and check its fields
+    # Create a taxonomy
+    When I request POST taxonomies with 9606 for ID, Homo Sapiens for name and TAXONOMY_1 for ancestors
```
I wonder where is TAXONOMY_1 is created?

selvaebi replied:
I guess there is no validation for ancestors so it passed, so you can make it NONE or create TAXONOMY_1 and refer to it

tskir replied:
You're right! Fixed the ancestor taxonomy to `NONE`

cyenyxe replied:
Why doesn't the application code fail when trying to use a non-existing ancestor taxonomy?

tskir replied:
Because this is the way it's currently implemented:

```
    @ManyToMany
    @JsonProperty
    private List<Taxonomy> ancestors;
```

There's no `@NotNull` constraint, hence the `ancestor` list is allowed to contain `null` values

tskir replied:
And we can't just trivially fix this because one "root" taxonomy entity must still be present

cyenyxe replied:
Right, I was thinking more on the lines of pointing to a taxonomy that does not exist. But I guess the model should control that.
